addons/mail/static/src/model/model_field.js

```diff
@@ -591,54 +608,83 @@ class ModelField {
      * @private
      * @param {mail.model} record
      * @param {mail.model|mail.model[]} newValue
+     * @param {Object} param2
+     * @param {boolean} param2.hasToUpdateInverse
      * @returns {boolean} whether the value changed for the current field
      */
-    _setRelationUnlinkX2Many(record, newValue) {
-        const toDeleteOtherRecords = this._setRelationConvertX2ManyValue(newValue);
+    _setRelationUnlinkX2Many(record, newValue, { hasToUpdateInverse }) {
+        const recordsToUnlink = this._convertX2ManyValue(
+            newValue,
+            { hasToVerify: false }
+        );
         const otherRecords = this.read(record);
 
-        let isDeleting = false;
-        for (const toDeleteOtherRecord of toDeleteOtherRecords) {
+        let hasChanged = false;
+        for (const recordToUnlink of recordsToUnlink) {
             // unlink other record from current record
-            const wasDeleted = otherRecords.delete(toDeleteOtherRecord);
-            if (!wasDeleted) {
+            const wasLinked = otherRecords.delete(recordToUnlink);
+            if (!wasLinked) {
                 continue;
             }
-            isDeleting = true;
+            hasChanged = true;
             // unlink current record from other records
-            toDeleteOtherRecord.update({
-                [this.inverse]: [['unlink', record]],
-            });
-            // apply causality
-            if (this.isCausal) {
-                toDeleteOtherRecord.delete();
+            if (hasToUpdateInverse) {
+                if (!recordToUnlink.exists()) {
+                    // This case should never happen ideally, but the current
+                    // way of handling related relational fields make it so that
+                    // deleted records are not always reflected immediately in
+                    // these related fields.
+                    continue;
+                }
+                this.env.modelManager._update(
+                    recordToUnlink,
+                    { [this.inverse]: [['unlink', record]] },
+                    { hasToUpdateInverse: false }
+                );
+                // apply causality
+                if (this.isCausal) {
+                    this.env.modelManager._delete(recordToUnlink);
+                }
             }
         }
-        return isDeleting;
+        return hasChanged;
     }
 
     /**
      * Handling of a `set` 'unlink' of a x2one relational field.
      *
      * @private
      * @param {mail.model} record
+     * @param {Object} param1
+     * @param {boolean} param1.hasToUpdateInverse
      * @returns {boolean} whether the value changed for the current field
      */
-    _setRelationUnlinkX2One(record) {
-        const toDeleteOtherRecord = this.read(record);
+    _setRelationUnlinkX2One(record, { hasToUpdateInverse }) {
+        const recordToUnlink = this.read(record);
```

Comment:
Why not just name it `previousOtherRecord` (or `prevOtherRecord`)? More declarative.